File: Python/tigre/utilities/io/DXChangeDataLoader.py
from __future__ import print_function
from __future__ import with_statement

import os
import math
import numpy as np
from tqdm import tqdm
import h5py

from configparser import ConfigParser
from tigre.utilities.geometry import geometry

def DXChangeDataLoader(filename):


    with h5py.File(filename, "r") as f:
        group = f['/exchange/']
        proj = group['data'][()]
        white = group['data_white'][()]
        dark = group['data_dark'][()]
        angles = group['theta'][()]*np.pi/180
    f.close()

    white = np.float32(white)
    dark = np.float32(dark)
    proj = np.float32(proj)

    white[white<0] = 0
    dark[dark<0] = 0
    dark = np.mean(dark,axis=0)
    white = np.mean(white,axis=0)
    proj[proj<=0]=0.001
    proj = (proj-dark)/(white-dark)
    proj = -np.log(proj/proj.max())
    geo = geometry(mode='parallel',nVoxel=np.array([proj.shape[1], proj.shape[2], proj.shape[2]]))

    return np.float32(proj), geo, angles

File: MATLAB/Utilities/IO/DXChange/DXChangeDataLoader.m
function [proj,geo,angles]=DXChangeDataLoader(fname)
%DXChangeDataLoader(filepath) Loads DXChange datasets into TIGRE standard
%
%   DXChangeDataLoader(fname OPT,VAL,...) uses options and values. 
%      These are options in case you don't want to load the entire
%      dataset, but only particular sets of projections. 

%%
% make sure its h5 file
[~, ~, fExt] = fileparts(fname);
if ~strcmp(fExt,'.h5')
    error('Filename is not .h5 file')
end
proj=single(h5read(fname,"/exchange/data/"));
white=single(h5read(fname,"/exchange/data_white/"));
dark=single(h5read(fname,"/exchange/data_dark/"));
white(white<0)=0;
dark(dark<0)=0;
white=mean(white,3);
dark=mean(dark,3);
dark=zeros(size(dark),"single");
proj=(proj-dark)./(white-dark);
proj(proj<0)=0.01;
proj=-log(proj/(max(proj(:))));
proj=permute(proj,[2 1 3]);


angles=h5read(fname,"/exchange/theta").'*pi/180;
geo=defaultGeometry('mode','parallel','nDetector',[size(proj,2),size(proj,1)]);

end